extensions/infinispan-cache/deployment/src/test/java/io/quarkus/cache/infinispan/InfinispanCacheTest.java

```diff
@@ -0,0 +1,339 @@
+        assertThat(remoteCache.get(id)).isNull();
+    }
+
+    @Test
+    public void testGetWithWithMaxidle() throws Exception {
+        Cache cache = getCache(-1, 3);
+        String id = generateId();
+        assertThat(awaitUni(cache.get(id, key -> "one"))).isEqualTo("one");
+        assertThat(awaitUni(cache.get(id, key -> "two"))).isEqualTo("one");
+        assertThat(remoteCache.get(id)).isEqualTo("one");
+        // Wait maxidle expiration
+        await().pollDelay(Duration.ofSeconds(3)).untilAsserted(() -> assertThat(true).isTrue());
+        // key has expired
+        assertThat(remoteCache.get(id)).isNull();
+        assertThat(awaitUni(cache.get(id, key -> "two"))).isEqualTo("two");
+        assertThat(remoteCache.get(id)).isEqualTo("two");
+        // Wait maxidle expiration
+        await().pollDelay(Duration.ofSeconds(3)).untilAsserted(() -> assertThat(true).isTrue());
+        assertThat(remoteCache.get(id)).isNull();
+    }
+
+    @Test
+    public void testGetWithNullValues() {
+        Cache cache = getCache();
+        String id = generateId();
+        assertThat(awaitUni(cache.get(id, key -> null))).isEqualTo(null);
+        assertThat(remoteCache.get(id)).isEqualTo(NullValue.NULL);
+    }
+
+    protected <T> Future<T> fork(Callable<T> c) {
+        return testExecutor.submit(new CallableWrapper<>(c));
+    }
+
+    private static class CallableWrapper<T> implements Callable<T> {
+        private final Callable<? extends T> c;
+
+        CallableWrapper(Callable<? extends T> c) {
+            this.c = c;
+        }
+
+        @Override
+        public T call() throws Exception {
+            try {
+                Log.trace("Started fork callable..");
+                T result = c.call();
+                Log.debug("Exiting fork callable.");
+                return result;
+            } catch (Exception e) {
+                Log.warn("Exiting fork callable due to exception", e);
+                throw e;
+            }
+        }
+    }
+
+    protected static ThreadFactory getTestThreadFactory(final String prefix) {
+        final String className = InfinispanCacheTest.class.getSimpleName();
+
+        return new ThreadFactory() {
+            private final AtomicInteger counter = new AtomicInteger(0);
+
+            @Override
+            public Thread newThread(Runnable r) {
+                String threadName = prefix + "-" + counter.incrementAndGet() + "," + className;
+                return new Thread(r, threadName);
+            }
+        };
+    }
+
+    @Test
+    public void testGetWithParallelCalls() throws Exception {
+        CyclicBarrier barrier = new CyclicBarrier(2);
+        Cache cache = getCache();
+        String id = generateId();
+        Future<Uni<String>> thread1 = fork(() -> cache.get(id, key -> {
+            try {
+                barrier.await(10, TimeUnit.SECONDS);
+                barrier.await(10, TimeUnit.SECONDS);
```

Comment:
why twice?

Reply:
this is a test coming from Spring integration in the infinispan repo. As explained by @jabolina one is to first make sure we are inside the lambda, for line 189. The second to wait inside the lambda until we issue the second request on line 192

Reply:
I would add that explanation in the code - because you can be sure I would ask the same question in 1 month... or one week, or even maybe tomorrow.